Working 3.10 install

TylerGubala opened the pull request:
Fixes some of the 3.10 code when creating an event (removed the `loop` parameter)

Adds some packages to the requirements file

TylerGubala committed:
Fix protobuf error

Protobuf was giving an error about being able to create descriptors directly.
By downgrading, we avoid the error.

TylerGubala committed:
Resolved Python 3.10 asyncio issue using loop parameter

TylerGubala committed:
Add PyOpenGL to dev requirements for 3D example

urastogi885 commented:
Thanks for the PR.
I tested the pull request. The changes definitely get rid of `loop` parameter error thrown when sending any requests to Vector via the SDK. However, half of the times the requests don't get executed with the following traceback:
```shell
vector-python-sdk % python3 examples/tutorials/01_hello_world.py
11:12:59.762     connection.Connection    INFO  Connecting to <VECTOR_IP>:443 for <VECTOR_NAME> using <VECTOR_CERT_PATH>
E0901 11:13:00.483381000 123145484791808 fork_posix.cc:76]             Other threads are currently calling into gRPC, skipping fork() handlers
11:13:00.597     connection.Connection    INFO  BehaviorControl control_granted_response { }
Traceback (most recent call last):
  File "anki_vector/connection.py", line 749, in log_handler
    result = await func(*args, **kwargs)
  File "anki_vector/animation.py", line 168, in load_animation_list
    return await self._load_animation_list()
  File "anki_vector/animation.py", line 134, in _load_animation_list
    result = await self.grpc_interface.ListAnimations(req)
  File "/usr/local/lib/python3.10/site-packages/aiogrpc/channel.py", line 40, in __call__
    return await fut
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.DEADLINE_EXCEEDED
        details = "ListAnimations request timed out"
        debug_error_string = "{"created":"@1662045186.236287000","description":"Error received from peer ipv4:<ROBOT_IP>:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"ListAnimations request timed out","grpc_status":4}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "examples/tutorials/01_hello_world.py", line 34, in <module>
    main()
  File "examples/tutorials/01_hello_world.py", line 28, in main
    with anki_vector.Robot(args.serial) as robot:
  File "anki_vector/robot.py", line 742, in __enter__
    self.connect(self.behavior_activation_timeout)
  File "anki_vector/robot.py", line 663, in connect
    anim_request = self._anim.load_animation_list()
  File "anki_vector/connection.py", line 806, in result
    return future.result()
  File "/usr/local/Cellar/python@3.10/3.10.6_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/_base.py", line 458, in result
    return self.__get_result()
  File "/usr/local/Cellar/python@3.10/3.10.6_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "anki_vector/connection.py", line 751, in log_handler
    raise connection_error(rpc_error) from rpc_error
anki_vector.exceptions.VectorTimeoutException: StatusCode.DEADLINE_EXCEEDED: ListAnimations request timed out

Message took too long to complete.
```
@TylerGubala did you experience similar issues?

TylerGubala commented:
That I did. Not exactly sure why.

Wanted to keep the pull request small in scope to start with, and hopefully
get the conversation started so I could familiarize myself with the code
base.

That said, can you point me in the right direction? Where can I start
looking in order to debug these timeout issues?

Thanks in advance!

On Thu, Sep 1, 2022, 11:39 AM Umang Rastogi ***@***.***>
wrote:

> Thanks for the changes suggested in the PR.
> I tested the pull request. The changes definitely get rid of loop
> parameter error thrown when sending any requests to Vector via the SDK.
> However, half of the times the requests don't get executed due to other
> errors with the following traceback:
>
> vector-python-sdk % python3 01_hello_world.py
> 11:12:59.762     connection.Connection    INFO  Connecting to <VECTOR_IP>:443 for <VECTOR_NAME> using <VECTOR_CERT_PATH>
> E0901 11:13:00.483381000 123145484791808 fork_posix.cc:76]             Other threads are currently calling into gRPC, skipping fork() handlers
> 11:13:00.597     connection.Connection    INFO  BehaviorControl control_granted_response { }
> Traceback (most recent call last):
>   File "vector-python-sdk/sdk/anki_vector/connection.py", line 749, in log_handler
>     result = await func(*args, **kwargs)
>   File "vector-python-sdk/sdk/anki_vector/animation.py", line 168, in load_animation_list
>     return await self._load_animation_list()
>   File "vector-python-sdk/sdk/anki_vector/animation.py", line 134, in _load_animation_list
>     result = await self.grpc_interface.ListAnimations(req)
>   File "/usr/local/lib/python3.10/site-packages/aiogrpc/channel.py", line 40, in __call__
>     return await fut
> grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
>         status = StatusCode.DEADLINE_EXCEEDED
>         details = "ListAnimations request timed out"
>         debug_error_string = ***@***.***","description":"Error received from peer ipv4:<ROBOT_IP>:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"ListAnimations request timed out","grpc_status":4}"
> >
>
> The above exception was the direct cause of the following exception:
>
> Traceback (most recent call last):
>   File "vector-python-sdk/sdk/examples/tutorials/01_hello_world.py", line 34, in <module>
>     main()
>   File "vector-python-sdk/sdk/examples/tutorials/01_hello_world.py", line 28, in main
>     with anki_vector.Robot(args.serial) as robot:
>   File "vector-python-sdk/sdk/anki_vector/robot.py", line 742, in __enter__
>     self.connect(self.behavior_activation_timeout)
>   File "vector-python-sdk/sdk/anki_vector/robot.py", line 663, in connect
>     anim_request = self._anim.load_animation_list()
>   File "vector-python-sdk/sdk/anki_vector/connection.py", line 806, in result
>     return future.result()
>   File ***@***.***/3.10.6_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/_base.py", line 458, in result
>     return self.__get_result()
>   File ***@***.***/3.10.6_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
>     raise self._exception
>   File "vector-python-sdk/sdk/anki_vector/connection.py", line 751, in log_handler
>     raise connection_error(rpc_error) from rpc_error
> anki_vector.exceptions.VectorTimeoutException: StatusCode.DEADLINE_EXCEEDED: ListAnimations request timed out
>
> Message took too long to complete.
>
> @TylerGubala <https://github.com/TylerGubala> did you experience similar
> issues?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/anki/vector-python-sdk/pull/52#issuecomment-1234454936>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFLEYR2SJXS3QR4XF6RE3OLV4DEZRANCNFSM6AAAAAAQASSAMQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

urastogi885 commented:
I think similar issues exist when using the SDK with Python 3.8 and 3.9
The problem lies in how the SDK uses the Python [Asyncio](https://docs.python.org/3/library/asyncio.html) library to execute concurrent tasks. There have been changes in the Asyncio library since Python 3.6 and we are still trying to adapt the SDK to those changes.

TylerGubala committed:
Ignore protobuf stuff in workspace

Wanted to work with the protobuf compiler in the workspace

TylerGubala committed:
Added VSCode debugging launch config

Makes it easier to run the current file and test

TylerGubala committed:
added protobuf rebuild task

Requires the following directory structure:

/google
   ... <google protobuf extra libs> ...
/.protobuf
   /bin
      protoc (protobuf exe)
   /include
      ... <protobuf standard libs>...

TylerGubala committed:
updated protobuf python objects

TylerGubala committed:
add asyncio gather to process Cancellation error

fixes #41

TylerGubala committed:
Updated reference to pillow image modes

urastogi885 commented:
In order to assess your PR, can you provide information about the protobuf version used and how did you update the pb2 files?

urastogi885 commented:
> Thanks for the PR. I tested the pull request. The changes definitely get rid of `loop` parameter error thrown when sending any requests to Vector via the SDK. However, half of the times the requests don't get executed with the following traceback:
> 
> ```shell
> vector-python-sdk % python3 examples/tutorials/01_hello_world.py
> 11:12:59.762     connection.Connection    INFO  Connecting to <VECTOR_IP>:443 for <VECTOR_NAME> using <VECTOR_CERT_PATH>
> E0901 11:13:00.483381000 123145484791808 fork_posix.cc:76]             Other threads are currently calling into gRPC, skipping fork() handlers
> 11:13:00.597     connection.Connection    INFO  BehaviorControl control_granted_response { }
> Traceback (most recent call last):
>   File "anki_vector/connection.py", line 749, in log_handler
>     result = await func(*args, **kwargs)
>   File "anki_vector/animation.py", line 168, in load_animation_list
>     return await self._load_animation_list()
>   File "anki_vector/animation.py", line 134, in _load_animation_list
>     result = await self.grpc_interface.ListAnimations(req)
>   File "/usr/local/lib/python3.10/site-packages/aiogrpc/channel.py", line 40, in __call__
>     return await fut
> grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
>         status = StatusCode.DEADLINE_EXCEEDED
>         details = "ListAnimations request timed out"
>         debug_error_string = "{"created":"@1662045186.236287000","description":"Error received from peer ipv4:<ROBOT_IP>:443","file":"src/core/lib/surface/call.cc","file_line":967,"grpc_message":"ListAnimations request timed out","grpc_status":4}"
> >
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "examples/tutorials/01_hello_world.py", line 34, in <module>
>     main()
>   File "examples/tutorials/01_hello_world.py", line 28, in main
>     with anki_vector.Robot(args.serial) as robot:
>   File "anki_vector/robot.py", line 742, in __enter__
>     self.connect(self.behavior_activation_timeout)
>   File "anki_vector/robot.py", line 663, in connect
>     anim_request = self._anim.load_animation_list()
>   File "anki_vector/connection.py", line 806, in result
>     return future.result()
>   File "/usr/local/Cellar/python@3.10/3.10.6_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/_base.py", line 458, in result
>     return self.__get_result()
>   File "/usr/local/Cellar/python@3.10/3.10.6_1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
>     raise self._exception
>   File "anki_vector/connection.py", line 751, in log_handler
>     raise connection_error(rpc_error) from rpc_error
> anki_vector.exceptions.VectorTimeoutException: StatusCode.DEADLINE_EXCEEDED: ListAnimations request timed out
> 
> Message took too long to complete.
> ```
> 
> @TylerGubala did you experience similar issues?

Also, @TylerGubala did the update of proto files reduce/remove above error thrown when executing SDK requests?

TylerGubala commented:
> In order to assess your PR, can you provide information about the protobuf version used

I ran the following command and got the output:

`py -m pip show protobuf`
> Name: protobuf
Version: 4.21.5
Summary: 
Home-page: https://developers.google.com/protocol-buffers/
Author: protobuf@googlegroups.com
Author-email: protobuf@googlegroups.com
License: 3-Clause BSD License
Location: d:\code\vector-python-sdk\venv\lib\site-packages
Requires: 
Required-by: googleapis-common-protos

---

> how did you update the pb2 files?

I downloaded `protoc` [here]()

![image](https://user-images.githubusercontent.com/22432839/189017840-538fb5cd-b4c6-48fc-9cca-b76939dc959c.png)

And placed the resulting `bin` and `include` files in the following location in the repo:

`.protobuf/`

![image](https://user-images.githubusercontent.com/22432839/189018491-61e8be8e-268c-4341-90ef-867493038cac.png)

Then I downloaded [the 'extra' Google protobuf libs](https://github.com/googleapis/googleapis) and placed the `google/` folder at the root of the repo like so:

![image](https://user-images.githubusercontent.com/22432839/189021970-37e9734a-04a9-48a4-a0da-4bf4b0f32eb1.png)

[This is apparently recommended practice](https://stackoverflow.com/a/66169476), though I don't love it.

Then in Visual Studio Code (my preferred IDE):

View -> Command Palette -> Tasks: Run Task -> `Build protobuf files`

![image](https://user-images.githubusercontent.com/22432839/189018071-ab4e3273-17fb-43a0-a6f5-476b22995c74.png)

Otherwise you can rebuild them from the terminal using the following command, assuming your current working directory is the root of the repo:

`.protobuf/bin/protoc --python_out=. anki_vector/messaging/*.proto`

---

> Also, @TylerGubala did the update of proto files reduce/remove above error thrown when executing SDK requests?

Nope I need help with that one. Not sure why it's happening suddenly. Maybe someone at DDL has a robot and can help with what the reason for that is and all the `events.EventHandler WARNING  Unknown Event type` that I keep getting.

TylerGubala committed:
Fix path to audio files in play audio example

Removes the need for the user to copy the audio files
This keeps their environment "cleaner"
